### PR TITLE
android: create persistent SharedPreferences-based KV store

### PR DESCRIPTION
**Pull request by goaway**
Description: Updates the exposed KeyValueStore type to be a more traditional implementable interface, and provides a simple persisting implementation based on Android SharedPreferences.
Risk: Low
Testing: Application

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by jpsim** [comment]
Looks straightforward enough. Should we expose a builder API to enable this?

**Comment by goaway** [comment]
@jpsim Setting a KV store is already supported in the Kotlin EngineBuilder, with Swift to follow. :)

File: library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt
Comment (goaway) [reply library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt:10]: Added docstring, and the interface is documented - would be great to generate docs from these.

**Comment by jpsim** [comment]
> Should we expose a builder API to enable this?
> > Setting a KV store is already supported in the Kotlin EngineBuilder

Sorry I wasn't clear. Right now to use this I think users would have to configure the builder like this:

```kotlin
.addKeyValueStore("some name the user has to choose", SharedPreferencesStore(...))
```

Why do we have to expose the `SharedPreferencesStore` in the public API at all? Could we not do this?

```kotlin
.useDefaultPersistentKeyValueStore(true)
```

And configure `SharedPreferencesStore` entirely internally to the library?

File: library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt
Comment (goaway) [reply library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt:10]: (Looks like @jpsim has made some great progress here!)

Comment (goaway) [reply library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt:10]: We don't today have a unit test setup that loads the Android platform APIs, because we don't today have a lot of Android-only or Android-specific code (by design). If/as we add more, we probably do want such a test suite, but this is mostly meant to be a trivial example implementation that guides others in using the API. That said, once we have a place for such tests to live, I agree we should cover this at the unit level for completeness' sake.

Comment (Augustyniak) [reply library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt:10]: I see, thank you for the context.

I would argue that we should have a test even for this trivial example since otherwise we cannot really tell whether it works or not the way we expect it to work - as in, I think that we are past the threshold when we can justify having an Android tests suite (if we have a need for Android code we have a need for Android tests). 

Could we create a GH issue for this if you do not want to do this as part of this PR? I would love us to start investing in testing more, otherwise it's going to become increasingly hard to continue to move forward without leaving broken stuff behind us.

Comment (jpsim) [reply library/kotlin/io/envoyproxy/envoymobile/android/SharedPreferencesStore.kt:10]: [Roboelectric](http://robolectric.org/) supports this today, and we use it in quite a few tests: https://github.com/envoyproxy/envoy-mobile/search?q=robolectric

**Comment by Augustyniak** [comment]
> Also, having reasonable/good defaults doesn't mean that we shouldn't expose ways to change them in case they want to refine/disable/change it for their use case.

That would be my preference as well -> make some defaults that work just fine for most users and allow them to override it when they need something custom. System level networking libraries on iOS/Android do not _require_ you to provide any caching configuration to get basic caching up and running.
 
If we want to experiment with this first that's fine but I would argue that the final API needs to have some notions of defaults implemented.

**Comment by goaway** [comment]
I think we're all in agreement here? The only thing I'm proposing is we hold off on making assumptions around what a default should be until we understand see a bit better how it's being used (alt-svc caching, http caching, etc.).

**Comment by jpsim** [comment]
> I think we're all in agreement here?

To recap some offline discussions, yes in agreement as a first version, which requires consumers to wire this through the `addKeyValueStore` builder API.

Ideally before merging this would still have the following:

1. changelog entry
2. doc updates
3. usage in the experimental Kotlin app

I still think a test using roboelectric to simulate Android's shared preferences would be good to have but understand if that's not something you want to tackle here.

**Review by Augustyniak — DISMISSED** [review]

File: docs/root/api/starting_envoy.rst
Comment (jpsim) [inline docs/root/api/starting_envoy.rst:443]: Missing the string parameter.

Maybe we should encourage reverse domain notation to convey that these should be unique?

```suggestion
  builder.addKeyValueStore("com.myapp.envoymobile.MyKeyValueStore", MyKeyValueStoreImpl())
```

File: docs/root/api/starting_envoy.rst
Comment (jpsim) [inline docs/root/api/starting_envoy.rst:437]: Nit:

```suggestion
made available to the library. General usage is supported, but typical future usage will be in
```

File: docs/root/api/starting_envoy.rst
Comment (jpsim) [inline docs/root/api/starting_envoy.rst:456]: Missing string parameter here too

**Review by jpsim — APPROVED** [review]
Nice work

File: docs/root/api/starting_envoy.rst
Comment (goaway) [reply docs/root/api/starting_envoy.rst:443]: thanks, good catch

**Review by Augustyniak — APPROVED** [review]

**Pull request merged**
